mobile-browser-based-version/server/src/logic/federated/handlers.js

```diff
@@ -0,0 +1,547 @@
+import path from 'path';
+import fs from 'fs';
+import msgpack from 'msgpack-lite';
+import * as config from '../../../server.config.js';
+import {
+  averageWeights,
+  assignWeightsToModel,
+} from '../../helpers/tfjs_helpers.js';
+import tasks from '../../tasks/tasks.js';
+import * as tf from '@tensorflow/tfjs';
+import '@tensorflow/tfjs-node';
+
+const REQUEST_TYPES = Object.freeze({
+  CONNECT: 'connect',
+  DISCONNECT: 'disconnect',
+  SELECTION_STATUS: 'selection-status',
+  AGGREGATION_STATUS: 'aggregation-status',
+  POST_WEIGHTS: 'post-weights',
+  GET_WEIGHTS: 'get-weights',
+  POST_SAMPLES: 'post-samples',
+  GET_SAMPLES: 'get-samples',
+  GET_TASKS: 'get-tasks',
+});
+/**
+ * Fraction of client gradients required on final round of federated training
+ * to proceed to the pooling step.
+ */
+const AGGREGATION_THRESHOLD = 0.8;
+/**
+ * Number of selected clients required to start federated training.
+ */
+const TRAINING_THRESHOLD = 2;
+/**
+ * Once a certain threshold has been hit, leave a small window for late clients
+ * to join the next federated training session.
```

Comment:
time unit?

Reply:
and before we go to fully async, we can keep the rounds-based scheme here, but should also allow for late joiners (after several rounds), and some users to drop out (still ok as the threshold can still be reached with the remaining ones)

Reply:
@giordano-lucas